File: ios/Sources/RynBridgeAnalytics/DefaultAnalyticsProvider.swift
import Foundation
import RynBridge

public final class DefaultAnalyticsProvider: AnalyticsProvider, @unchecked Sendable {
    private let queue = DispatchQueue(label: "io.rynbridge.analytics")
    private var events: [(name: String, params: [String: AnyCodable], timestamp: Date)] = []
    private var userProperties: [String: String] = [:]
    private var userId: String?
    private var currentScreen: String?
    private var enabled: Bool = true

    public init() {}

    public func logEvent(name: String, params: [String: AnyCodable]) {
        queue.sync {
            guard enabled else { return }
            events.append((name: name, params: params, timestamp: Date()))
        }
    }

    public func setUserProperty(key: String, value: String) {
        queue.sync {
            userProperties[key] = value
        }
    }

    public func setUserId(_ userId: String) {
        queue.sync {
            self.userId = userId
        }
    }

    public func setScreen(name: String) {
        queue.sync {
            self.currentScreen = name
        }
    }

    public func resetUser() {
        queue.sync {
            userId = nil
            userProperties.removeAll()
        }
    }

    public func setEnabled(_ enabled: Bool) async throws -> Bool {
        queue.sync {
            self.enabled = enabled
        }
        return enabled
    }

    public func isEnabled() async throws -> Bool {
        return queue.sync { enabled }
    }
}

File: android/analytics/src/main/kotlin/com/devryner/rynbridge/analytics/DefaultAnalyticsProvider.kt
package com.devryner.rynbridge.analytics

import com.devryner.rynbridge.core.BridgeValue
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

class DefaultAnalyticsProvider : AnalyticsProvider {

    private val events = CopyOnWriteArrayList<Map<String, Any?>>()
    private val userProperties = ConcurrentHashMap<String, String>()
    @Volatile private var userId: String? = null
    @Volatile private var currentScreen: String? = null
    @Volatile private var enabled: Boolean = true

    override fun logEvent(name: String, params: Map<String, BridgeValue>) {
        if (!enabled) return
        events.add(
            mapOf(
                "name" to name,
                "params" to params,
                "screen" to currentScreen,
                "userId" to userId,
                "timestamp" to System.currentTimeMillis()
            )
        )
    }

    override fun setUserProperty(key: String, value: String) {
        userProperties[key] = value
    }

    override fun setUserId(userId: String) {
        this.userId = userId
    }

    override fun setScreen(name: String) {
        this.currentScreen = name
    }

    override fun resetUser() {
        userId = null
        currentScreen = null
        userProperties.clear()
    }

    override suspend fun setEnabled(enabled: Boolean): Boolean {
        this.enabled = enabled
        return this.enabled
    }

    override suspend fun isEnabled(): Boolean {
        return enabled
    }
}
